src/include/daos_pipeline.h

```diff
@@ -0,0 +1,261 @@
+	 *          DAOS_FILTER_FUNC_LE:		<=
+	 *          DAOS_FILTER_FUNC_GE:		>=
+	 *          DAOS_FILTER_FUNC_GT:		>
+	 *          DAOS_FILTER_FUNC_LIKE:		== (reg exp.)
+	 *          DAOS_FILTER_FUNC_ISNULL:		==NULL
+	 *          DAOS_FILTER_FUNC_ISNOTNULL:		!=NULL
+	 *          DAOS_FILTER_FUNC_AND:		&&
+	 *          DAOS_FILTER_FUNC_OR:		||
+	 *      - aggeration functions:
+	 *          DAOS_FILTER_FUNC_SUM:		SUM()
+	 *          DAOS_FILTER_FUNC_MIN:		MIN()
+	 *          DAOS_FILTER_FUNC_MAX:		MAX()
+	 *          DAOS_FILTER_FUNC_AVG:		AVG()
+	 *   -- key:
+	 *          DAOS_FILTER_DKEY:	Filter object represents dkey
+	 *          DAOS_FILTER_AKEY	Filter object represents akey
+	 *   -- constant:
+	 *          DAOS_FILTER_CONST:	Filter object is a constant
+	 */
+	int		filter_type;
+	/**
+	 * Type of data. Only relevant for keys and constant filter type objects:
+	 *          DAOS_FILTER_TYPE_BINARY
+	 *          DAOS_FILTER_TYPE_STRING
+	 *          DAOS_FILTER_TYPE_INTEGER
+	 *          DAOS_FILTER_TYPE_REAL
+	 */
+	int		data_type;
+	/**
+	 * Number of parameters for this filter object. For example, for '=='
+	 * we have 2 parameters.
+	 */
+	uint32_t	num_params;
+	/**
+	 * If filtering by akey, this tells us which one.
+	 */
+	d_iov_t		akey;
+	/**
+	 * How many constants we have in \a constant
+	 */
+	size_t		num_constants;
+	/**
+	 * This object holds the value of the constants
+	 */
+	d_iov_t		*constant;
+	/**
+	 * If filter should only be applied starting at an offset of the data.
+	 */
+	size_t		data_offset;
+	/**
+	 * Size of the data to be filtered.
+	 */
+	size_t		data_len;
+} daos_pipeline_filter_t;
+
+/**
+ * A pipeline node, used to build a pipeline.
+ */
+typedef struct {
+	/**
+	 * Node type can be any of the following:
+	 *   -- DAOS_PIPELINE_CONDITION:
+	 *          Records in, and records (meeting condition) out
+	 *   -- DAOS_PIPELINE_AGGREGATION:
+	 *          Records in, a single value out
+	 *
+	 * NOTE: Pipeline nodes can only be chained the following way:
+	 *             (condition) --> (condition)
+	 *             (condition) --> (aggregation)
+	 */
+	int			node_type;
+	/**
+	 * Number of filters inside this pipeline node
+	 */
+	size_t			num_filters;
+	/**
+	 * Array of filters for this node
+	 */
+	daos_pipeline_filter_t	*filters;
+} daos_pipeline_node_t;
+
+/**
+ * A pipeline.
+ */
+typedef struct {
+	/**
+	 * Version number of the data structure.
+	 */
+	uint64_t		version;
+	/**
+	 * Number of nodes chained in this pipeline
+	 */
+	size_t			num_nodes;
+	/**
+	 * Array of nodes for this pipeline
+	 */
+	daos_pipeline_node_t	*nodes;
+} daos_pipeline_t;
+
+
+/**
+ * Adds a new pipeline node to the pipeline 'pipeline' object. The effect of
+ * this function is to "push back" the new node at the end of the pipeline.
+ *
+ * \param[in,out]	pipeline	Pipeline object.
+ *
+ * \param[in]		node		Node object to be added to the pipeline.
+*/
+int
+daos_pipeline_push(daos_pipeline_t *pipeline, daos_pipeline_node_t *node);
```

Comment:
I would call this add and use push when we push it to the server for execution.

Reply:
Actually, @mchaarawi and I just have a discussion about this, and we believe that it is better to use "run" for execution than "push". Push makes it look as if the server is state-full. That is, you can "push" the pipeline and then execute it multiple times. But it is better to go with the overall design philosophy of DAOS and make this API state-less, which means that "run" fits better here (IMHO), because we pass the pipeline object every time and that triggers the run.

We can surely change push to add for the other helper functions that build the pipeline.